src/server.c

```diff
@@ -674,6 +675,17 @@ void tryResizeHashTables(int dbid) {
             db->sub_dict[subdict].resize_cursor = dbGetNextNonEmptySlot(db, slot, subdict);
         }
     }
+    /* Slow resize: try to resize all dicts to recycle memory. */
+    for (dbKeyType subdict = DB_MAIN; subdict <= DB_EXPIRES; subdict++) {
+        for (int i = 0; i < CRON_DBS_PER_CALL && i < db->dict_count; i++) {
+            int slot = db->sub_dict[subdict].slow_resize_cursor;
+            dict *d = subdict == DB_MAIN ? db->dict[slot] : db->expires[slot];
+            if (htNeedsResize(d)) {
```

Comment:
maybe we should skip non-empty dicts? (covered by the other loop)
thinking of cluster mode disabled, we process the one dict twice for nothing.

Reply:
If it is already resized, it should get skipped here. Don't see a issue with this logic.

But if we want to be explicit that slow resize is for empty dictionary, maybe we could rename the cursor to `empty_dict_resize_cursor` ?

Reply:
It is not just for empty dicts, but for all dicts. 

In cluster modes, the two cursors traverse over the dict array in different speed, so an extreme case is that a dict is empty when the cursor for non-empty dicts arrives, and then after serveral rounds, it becomes non-empty when the crusor for empty dicts arrives. In this case the dict would be missed. 

So the "slow" cursor needs to handle all dicts.

Reply:
i don't understand the concept of slow and fast here.
let's imagine both scan the same type (empty or non-empty doesn't matter), is there still some benefit of having both slow and fast scan?
what about cluster-mode-disabled? why do we need both?

Reply:
Perhaps the dict needs a buffer time? The dict may shrink and expand repeatedly if there is only keys being added and deleted?

Reply:
@oranagra 
However it does make sense to automaticly shrink the dict when keys become sparse, since the we already make the dict expand itself when in need. We may need a proper threshold to avoid constantly shrinking and expanding. In this case, we no longer need the "fast" scan.

Reply:
Automatic shrinking during `dictDelete` is reasonable, but I believe `tryResizeHashTables` should still be retained.

There is a scenario where batch data import triggers `bgsave` or `bgrewriteaof`, and both automatic expand and shrink are paused due to the presence of the child process. If the data import completes before the child process finishes, subsequent operations are only reads. In this case, the main database may remain either too dense or too sparse. Therefore, it is still necessary to use cron to periodically check and perform expand or shrink operations.

Reply:
I'm not aware of the history here.

Reply:
discussion will continue in #12850

Reply:
i'm having hard time to accept the slow and fast approach here (seemingly unrelated to empty and non-empty dicts), and the fact that on a common single-slot, single-db instance, we process that dict twice.
not sure how to proceed. maybe the approach suggested here is the lesser evil.
maybe we need a dict iterator that iterates over owned slots / dicts, rather than non-empty ones.
@yossigo maybe you can advise?